### PR TITLE
Add detachObserver

### PR DESCRIPTION
**Pull request by alex7r**
#### Summary of Changes

Adding detach function for observers
#### Testing Instructions

in Class extends JTable
use
$this->_observers->detachObserver('specObserver');

Adds ability to detach observers

### CODE REVIEW

**Comment by wojsmol** [comment]
@alex7r Please see https://github.com/alex7r/joomla-cms/pull/3

**Comment by wilsonge** [comment]
What's your use case for this. I'm hugely nervous of extensions disabling other extensions...

**Comment by alex7r** [comment]
For example I have an object that can have one role or another:
User can be client, team member or team leader.
Role functionality implemented by observer because we know which role user have only depending on data that were loaded to user object.
We don't want to pull everything once again from data base and create new object, we can attach observer that corresponds to user role.
It's all working great.

But if team member becoming leader we need to apply leader observer instead of member. So detaching observer would be nice.

Also I don't quite understand your vision on this, as I attach observer to JTableExtender instance and how can I access same instance from other extension? JTable::getInstance always returns new Class()

**Comment by alex7r** [comment]
@wilsonge don't you agree, would it be merged?

**Comment by ggppdk** [comment]
@alex7r 

if you want to detach observers added by **your own extensions**, 
- then this should not be needed ?

because you can modify the code of the observer itself, to achieve what you need to do, right ?
even if your code looks less "clean"

**Comment by alex7r** [comment]
Sorry, but what you mean by this?
Can you provide working example which would not be hack for making something that should be essential?
____
I'll write a bit mess, but I believe everybody will get the point:

`JTable{`
`public function attachObserver(JObserverInterface $observer)`
`	{`
`		$this->_observers->attachObserver($observer);`
`	}`
`}`

`TUser Object`
`(`
`    [_tbl:protected] => #__users`
`    [_observers:protected] => JObserverUpdater Object`
`        (`
`            [observers:protected] => Array`
`                (`
`                    [UserObserver] => UserObserver Object`
`                        (`
`                            [table:protected] => TUser Object`
` *RECURSION*`
`                        )`
`                    [TUserObserverCoordinator] => TUserObserverCoordinator Object`
`                        (`
`                            [table:protected] => TUser Object`
` *RECURSION*`
`                        )`
`                )`
`            [doCallObservers:protected] => 1`
`        )`
`)`
____
So as you can see from all the above:
I can access `$this->_observers`
But I can't access `$this->_observers->observers` and iterate and unset as this is protected

**Comment by alex7r** [comment]
@ggppdk you should be solid with

> even if your code looks less "clean"

in this  situation means 

> even if you reinvent the wheel

**Comment by ggppdk** [comment]
@alex7r 

About accepting this feature, it is not my call, 
what you need to do for this to be considered, is :

- prove / explain it is useful **and common** enough (i think you tried / done this already ?) 
- convince maintainers it will not be misused or cause problems, **that when such problems appear they will not have to run for it, to fix things**

and usually "new feature" PRs wait for much longer time that "bug-fixes" PRs, i hope you are not disappointed with it, if you make a few 'new feature' PRs, eventually one of them will be accepted

again this is not my call

**Comment by alex7r** [comment]
**Just to reference my points once again:**

https://github.com/joomla/joomla-cms/pull/11037#issuecomment-231337114
- [x] prove / explain it is useful and common enough

https://github.com/joomla/joomla-cms/pull/11037#issuecomment-231052752
- [x] convince maintainers it will not be misused or cause problems

____

@ggppdk , well I know and understand that it's not your call. I'm just don't letting you to impose fear and doubts on call-makers.

**Comment by ggppdk** [comment]
>  I'm just don't letting you to impose fear and doubts on call-makers.

you over-estimate my effects , anyway i understand

**Comment by alex7r** [comment]
I believe if any one says that something like

> It's not compatible

or

> Well, is it needed?

it is a ring call for rechecking and doubting changes. And I also think that it's right position in terms of repo maintaining as millions are using this code and

> with great power come great responsibility


So some explanation like why this way and not another and why it's safe is good thing.

**Comment by wilsonge** [comment]
Ok I think the use case is fine. Does this patch work tho? Just on code review (as on holiday) isn't the key the class name rather than the observer itself?

**Comment by alex7r** [comment]
@wilsonge , yes it works.
May be I should have change var name, but I thought that 'key' is not the best Idea, 'classname' doesn't suite well, so I leaved 'observer'
But if you'll look at PHPDoc - @param String 
and described as 'observer class name'

File: libraries/joomla/observer/updater.php [hunk 1]
```diff
@@ -60,6 +60,24 @@ public function attachObserver(JObserverInterface $observer)
...
+	 *
+	 * @param   String  $observer  The observer class name
+	 *
+	 * @return  void
+	 *
+	 * @since   3.6.0
+	 */
+	public function detachObserver($observer)
+	{
+		if (isset($this->observers[$observer]))
```
Comment (bertmert) [inline libraries/joomla/observer/updater.php]: You don't need to check that a variable is set before you unset it.

Comment (alex7r) [inline libraries/joomla/observer/updater.php]: Well, AFAIK unsetting not set index will give you **index not exists notice**.
So why would you trigger notices if you could avoid it? )

**Comment by alex7r** [comment]
@wilsonge , @mbabker 
Can any one with merge permissions take a lot at it once again?

File: libraries/joomla/observer/updater.php (hunk 1 above)
Comment (bertmert) [inline libraries/joomla/observer/updater.php]: In this case you would get a `Notice: Undefined variable: $this->observers` if $this->observers would not be defined. But it's always defined as array. A not existing index doesn't throw a Notice.

Comment (alex7r) [inline libraries/joomla/observer/updater.php]: did you test this behavior, or just suggesting?

**Comment by wilsonge** [comment]
Yes - but i'm holding on merging things until we're satisfied with 3.6.0 and are sure we don't need to make any quick follow up releases

File: libraries/joomla/observer/updater.php (hunk 1 above)
Comment (bertmert) [inline libraries/joomla/observer/updater.php]: All I suggest I test before.

**Comment by alex7r** [comment]
3.6.0 is out, Still waiting )

**Comment by alex7r** [comment]
Some status updates? @wilsonge

**Comment by brianteeman** [comment]
It looks to me that @wilsonge marked this as RTC but he did it wrong and the cms-bot removed the label. I will set it back to RTC now - it would not have gone in a 3.6.x release anyway as it is not a bug fix.

**Comment by brianteeman** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11037">issues.joomla.org/tracker/joomla-cms/11037</a>.</sub>

**Comment by alex7r** [comment]
thanks

**Pull request merged**